### PR TITLE
User joining existing game

### PR DESCRIPTION
**Pull request by wariofdz**
## Changed: 

- Added new enum `JoinExistingGameStatus.ts`.
- Added new interface `UseJoinExistingGameResult.ts`.
- Added new component `JoinExistingGame.tsx` and his test.
- Added new hook `useJoinExistingGame.ts` and his test.
- Added new style `_page.scss` to `JoinExistingGame` component.
- Modify file `GameRoutes.tsx` to add new route to join a game.
- Modify file `_page.scss` to set the maximum div in the pending and active games.
- Modify `Login` to include redirecTo scenario.

### CODE REVIEW

File: packages/frontend/web-ui/src/auth/pages/Login.tsx [hunk 1]
```diff
@@ -28,6 +28,11 @@ import { useLoginForm } from '../hooks/useLoginForm';
 import { LoginStatus } from '../models/LoginStatus';
 import { UseLoginFormResult } from '../models/UseLoginFormResult';
 
+export interface StateType {
+  state: { key: string };
+  search: string;
+}
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/auth/pages/Login.tsx:34]: What is the purpose of this interface?

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            break;
+          } else {
+            if (auth === null) {
+              navigate(buildLoginPageUrl(url.toString()));
+            }
+            if (user === null && auth !== null) {
+              await dispatch(getUserMe(auth.accessToken));
+            }
+          }
+          setStatus(JoinExistingGameStatus.pending);
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.ts:57]:
```suggestion
          } else {
            if (auth === null) {
              navigate(buildLoginPageUrl(url.toString()));
            }
            if (user === null && auth !== null) {
              await dispatch(getUserMe(auth.accessToken));
            }
            setStatus(JoinExistingGameStatus.pending);
          }
```

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 3]
```diff
@@ -0,0 +1,792 @@
...
+
+        (
+          useAppSelector as unknown as jest.Mock<typeof useAppSelector>
+        ).mockReturnValueOnce(userFixture);
+
+        (useJoinGame as jest.Mock<typeof useJoinGame>).mockReturnValueOnce(
+          singleApiCallHookResultFixture,
+        );
+
+        await waitFor(() => {
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts:304]: Why shouldn't we use `act` instead?

**Review by notaphplover — CHANGES_REQUESTED** [review]
Nice, we should fix some minor issues though

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts (hunk 3 above)
Comment (wariofdz) [reply packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts:304]: Finally, I put `act` with `async` word inside to allow that all of promises has been resolved. The warning is not appear.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 4]
```diff
@@ -0,0 +1,713 @@
...
+          navigateMock,
+        );
+
+        dispatchMock = jest
+          .fn<ReturnType<typeof useAppDispatch>>()
+          .mockImplementationOnce(
+            <TReturn, TAction>(): TAction | TReturn =>
+              payloadActionFixture as TReturn,
+          ) as ReturnType<typeof useAppDispatch> &
+          jest.Mock<ReturnType<typeof useAppDispatch>>;
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts:135]: Let's not implement a mock which is not called in the flow.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 5]
```diff
@@ -0,0 +1,713 @@
...
+          .mockReturnValueOnce(singleApiCallHookResultFixture);
+
+        (
+          getUserMe as unknown as jest.Mock<typeof getUserMe>
+        ).mockReturnValueOnce(getUserMeResult);
+
+        navigateMock = jest
+          .fn<ReturnType<typeof useNavigate>>()
+          .mockReturnValue(undefined) as ReturnType<typeof useNavigate> &
+          jest.Mock<ReturnType<typeof useNavigate>>;
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts:228]: Let's not implement a mock which is not called in the flow.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 6]
```diff
@@ -0,0 +1,713 @@
...
+          navigateMock,
+        );
+
+        dispatchMock = jest
+          .fn<ReturnType<typeof useAppDispatch>>()
+          .mockImplementationOnce(
+            <TReturn, TAction>(): TAction | TReturn =>
+              payloadActionFixture as TReturn,
+          ) as ReturnType<typeof useAppDispatch> &
+          jest.Mock<ReturnType<typeof useAppDispatch>>;
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts:361]: Let's not implement a mock which is not called in the flow.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 7]
```diff
@@ -0,0 +1,713 @@
...
+          .mockReturnValueOnce(singleApiCallHookResultFixture);
+
+        (
+          getUserMe as unknown as jest.Mock<typeof getUserMe>
+        ).mockReturnValueOnce(getUserMeResult);
+
+        navigateMock = jest
+          .fn<ReturnType<typeof useNavigate>>()
+          .mockReturnValue(undefined) as ReturnType<typeof useNavigate> &
+          jest.Mock<ReturnType<typeof useNavigate>>;
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts:349]: Let's not implement a mock which is not called in the flow.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+          navigateMock,
+        );
+
+        dispatchMock = jest
+          .fn<ReturnType<typeof useAppDispatch>>()
+          .mockImplementationOnce(
+            <TReturn, TAction>(): TAction | TReturn =>
+              payloadActionFixture as TReturn,
+          ) as ReturnType<typeof useAppDispatch> &
+          jest.Mock<ReturnType<typeof useAppDispatch>>;
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts:512]: Let's not implement a mock which is not called in the flow

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 9]
```diff
@@ -0,0 +1,713 @@
...
+          .mockReturnValueOnce(singleApiCallHookResultFixture);
+
+        (
+          getUserMe as unknown as jest.Mock<typeof getUserMe>
+        ).mockReturnValueOnce(getUserMeResult);
+
+        navigateMock = jest
+          .fn<ReturnType<typeof useNavigate>>()
+          .mockReturnValue(undefined) as ReturnType<typeof useNavigate> &
+          jest.Mock<ReturnType<typeof useNavigate>>;
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts:500]: Let's not implement a mock which is not called in the flow.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 10]
```diff
@@ -0,0 +1,713 @@
...
+          navigateMock,
+        );
+
+        dispatchMock = jest
+          .fn<ReturnType<typeof useAppDispatch>>()
+          .mockImplementationOnce(
+            <TReturn, TAction>(): TAction | TReturn =>
+              payloadActionFixture as TReturn,
+          ) as ReturnType<typeof useAppDispatch> &
+          jest.Mock<ReturnType<typeof useAppDispatch>>;
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts:657]: Let's not implement a mock which is not called in the flow.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 11]
```diff
@@ -0,0 +1,713 @@
...
+          .mockReturnValueOnce(singleApiCallHookResultFixture);
+
+        (
+          getUserMe as unknown as jest.Mock<typeof getUserMe>
+        ).mockReturnValueOnce(getUserMeResult);
+
+        navigateMock = jest
+          .fn<ReturnType<typeof useNavigate>>()
+          .mockReturnValue(undefined) as ReturnType<typeof useNavigate> &
+          jest.Mock<ReturnType<typeof useNavigate>>;
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts:645]: Let's not implement a mock which is not called in the flow.

**Review by notaphplover — CHANGES_REQUESTED** [review]
We're almost there :), let's polish the test and merge it

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 12]
```diff
@@ -149,10 +129,6 @@ describe(useJoinExistingGame.name, () => {
         jest.clearAllMocks();
       });
 
-      it('should not call useDispatch()', () => {
-        expect(dispatchMock).not.toHaveBeenCalled();
-      });
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts]: Let's add the missing assertion.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 13]
```diff
@@ -273,10 +244,6 @@ describe(useJoinExistingGame.name, () => {
         expect(useJoinGame).toHaveBeenNthCalledWith(2);
         expect(useJoinGame).toHaveBeenNthCalledWith(3);
       });
-
-      it('should not call navigate()', () => {
-        expect(navigateMock).not.toHaveBeenCalled();
-      });
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts]: Let's add the missing assertion.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 14]
```diff
@@ -402,10 +344,6 @@ describe(useJoinExistingGame.name, () => {
         jest.clearAllMocks();
       });
 
-      it('should not call useAppDispatch()', () => {
-        expect(dispatchMock).not.toHaveBeenCalled();
-      });
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts]: Let's add the missing assertion.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 15]
```diff
@@ -419,10 +357,6 @@ describe(useJoinExistingGame.name, () => {
         expect(useJoinGame).toHaveBeenNthCalledWith(4);
       });
 
-      it('should not call navigate()', () => {
-        expect(navigateMock).not.toHaveBeenCalled();
-      });
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts]: Let's add the missing assertion.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 16]
```diff
@@ -529,10 +438,6 @@ describe(useJoinExistingGame.name, () => {
         jest.clearAllMocks();
       });
 
-      it('should not call useAppDispatch()', () => {
-        expect(dispatchMock).not.toHaveBeenCalled();
-      });
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts]: Let's add the missing assertion.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 17]
```diff
@@ -546,10 +451,6 @@ describe(useJoinExistingGame.name, () => {
         expect(useJoinGame).toHaveBeenNthCalledWith(4);
       });
 
-      it('should not call navigate()', () => {
-        expect(navigateMock).not.toHaveBeenCalled();
-      });
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts]: Let's add the missing assertion.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 18]
```diff
@@ -674,10 +550,6 @@ describe(useJoinExistingGame.name, () => {
         jest.clearAllMocks();
       });
 
-      it('should not call useAppDispatch()', () => {
-        expect(dispatchMock).not.toHaveBeenCalled();
-      });
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts]: Let's add the missing assertion.

File: packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts [hunk 19]
```diff
@@ -689,10 +561,6 @@ describe(useJoinExistingGame.name, () => {
         expect(useJoinGame).toHaveBeenNthCalledWith(2);
       });
 
-      it('should not call navigate()', () => {
-        expect(navigateMock).not.toHaveBeenCalled();
-      });
```
Comment (notaphplover) [inline packages/frontend/web-ui/src/game/hooks/useJoinExistingGame.spec.ts]: Let's add the missing assertion.

**Review by notaphplover — CHANGES_REQUESTED** [review]
Let's add all the missing assertions.

**Review by notaphplover — APPROVED** [review]

**Pull request merged**